flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/SlotSharingManager.java

```diff
@@ -0,0 +1,722 @@
+			return logicalSlotFuture;
+		}
+
+		@Override
+		public boolean release(Throwable cause) {
+			logicalSlotFuture.completeExceptionally(cause);
+
+			boolean pendingLogicalSlotRelease = false;
+
+			if (logicalSlotFuture.isDone() && !logicalSlotFuture.isCompletedExceptionally()) {
+				// we have a single task slot which we first have to release
+				final LogicalSlot logicalSlot = logicalSlotFuture.getNow(null);
+
+				if (logicalSlot != null && logicalSlot.isAlive()) {
+					pendingLogicalSlotRelease = logicalSlot.releaseSlot(cause).isDone();
+				}
+			}
+
+			if (!pendingLogicalSlotRelease) {
+				parent.releaseChild(getGroupId());
+			}
+
+			return !pendingLogicalSlotRelease;
+		}
+	}
+
+	// ------------------------------------------------------------------------
+	// Methods and classes for testing
+	// ------------------------------------------------------------------------
+
+	/**
+	 * Returns a collection of all resolved root slots.
+	 *
+	 * @return Collection of all resolved root slots
+	 */
+	@VisibleForTesting
+	public Collection<MultiTaskSlot> getResolvedRootSlots() {
+		ResolvedRootSlotValues vs = resolvedMultiTaskSlotValues;
+
+		if (vs == null ){
+			vs = new ResolvedRootSlotValues();
+			resolvedMultiTaskSlotValues = vs;
+		}
+
+		return vs;
+	}
+
+	@VisibleForTesting
+	Collection<MultiTaskSlot> getUnresolvedRootSlots() {
+		return unresolvedRootSlots.values();
+	}
+
+	/**
+	 * Collection of all resolved {@link MultiTaskSlot} root slots.
+	 */
+	private final class ResolvedRootSlotValues extends AbstractCollection<MultiTaskSlot> {
+
+		@Override
+		public Iterator<MultiTaskSlot> iterator() {
+			return new ResolvedRootSlotIterator(resolvedRootSlots.values().iterator());
+		}
+
+		@Override
+		public int size() {
+			int numberResolvedMultiTaskSlots = 0;
+
+			for (Set<MultiTaskSlot> multiTaskSlots : resolvedRootSlots.values()) {
+				numberResolvedMultiTaskSlots += multiTaskSlots.size();
+			}
+
+			return numberResolvedMultiTaskSlots;
+		}
+	}
+
+	/**
+	 * Iterator over all resolved {@link MultiTaskSlot} root slots.
+	 */
+	private static final class ResolvedRootSlotIterator implements Iterator<MultiTaskSlot> {
+		private final Iterator<Set<MultiTaskSlot>> baseIterator;
+		private Iterator<MultiTaskSlot> currentIterator;
+
+		private ResolvedRootSlotIterator(Iterator<Set<MultiTaskSlot>> baseIterator) {
+			this.baseIterator = Preconditions.checkNotNull(baseIterator);
+
+			if (baseIterator.hasNext()) {
+				currentIterator = baseIterator.next().iterator();
+			} else {
+				currentIterator = Collections.emptyIterator();
+			}
+		}
+
+		@Override
+		public boolean hasNext() {
+			progressToNextElement();
+
+			return currentIterator.hasNext();
+		}
+
+		@Override
+		public MultiTaskSlot next() {
+			progressToNextElement();
+
+			return currentIterator.next();
+		}
+
+		private void progressToNextElement() {
+			while(baseIterator.hasNext() && ! currentIterator.hasNext()) {
```

Comment:
nit: missing space between `while` and `(`.  Space after `!`.

Reply:
Good catch. Will correct it.